### PR TITLE
cmd/bosun: Add metric stats to host API

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

File: cmd/bosun/sched/host.go [hunk 1]
```diff
@@ -0,0 +1,809 @@
+package sched // import "bosun.org/cmd/bosun/sched"
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: thanks for making a new file :)

File: cmd/bosun/sched/host.go [hunk 2]
```diff
@@ -0,0 +1,809 @@
...
+	}
+	for name, host := range hosts {
+		host.Name = name
+		hostTagSet := opentsdb.TagSet{"host": host.Name}
+		hostMetadata, err := s.GetMetadata("", hostTagSet)
+		if err != nil {
+			slog.Error(err)
+		}
+		processHostIncidents(host, states, silences)
+		for _, ts := range icmpTimeOutTags {
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: this is a lot of iteration. I each iteration over hosts, you iterate over all tags for a lot of metrics. Perhaps we can turn those arrays into maps by host before we enter the main loop.

File: cmd/bosun/sched/host.go [hunk 3]
```diff
@@ -0,0 +1,809 @@
...
+				continue
+			}
+			// The host tag represents the polling source for these set of metrics
+			source, ok := ts["host"]
+			if !ok {
+				slog.Errorf("couldn't find source tag for icmp data for host %s", host.Name)
+			}
+			// 1 Means it timed out
+			timeout, timestamp, err := s.Search.GetLast("bosun.ping.timeout", ts.String(), false)
+			if err != nil && !(timestamp > 0) {
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: I think you mean `err != nil || timestamp <= 0`

File: cmd/bosun/sched/host.go [hunk 4]
```diff
@@ -0,0 +1,809 @@
...
+		}
+		for _, ts := range hwTempsTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			name, ok := ts["name"]
+			if !ok {
+				slog.Errorf("couldn't find name tag %s for host %s", host.Name, name)
+			}
+			tStatus, timestamp, err := s.Search.GetLast("hw.chassis.temps", ts.String(), false)
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: I notice you almost never check errors on these calls. Are you ok with zero values? Use _ if you aren't going to care about a particular error.

File: cmd/bosun/sched/host.go [hunk 5]
```diff
@@ -0,0 +1,809 @@
...
+				if m.Time.Before(time.Now().Add(-timeFilterAge)) {
+					continue
+				}
+				if !m.Tags.Equal(ts) {
+					continue
+				}
+				switch val := m.Value.(type) {
+				case string:
+					switch m.Name {
+					case "psMeta":
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: maybe check for name = psMeta earlier in the metadata loop. It should be cheaper than tagset equals.

File: cmd/bosun/sched/host.go [hunk 6]
```diff
@@ -0,0 +1,809 @@
...
+				status = "Ok"
+			}
+			if err == nil && timestamp > 0 {
+				host.Hardware.Storage.Batteries[id] = &Battery{
+					Status:            status,
+					StatusLastUpdated: timestamp,
+				}
+			}
+		}
+		for _, ts := range hwBoardPowerTags {
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: if you make these maps then these loops become `for _,ts := range hwBoardPowerTags[host.Name]` and the if/continue is unnecessary.

File: cmd/bosun/sched/host.go [hunk 7]
```diff
@@ -0,0 +1,809 @@
...
+			pd := &PhysicalDisk{}
+			if err == nil && timestamp > 0 {
+				pd.Status = status
+				pd.StatusLastUpdated = timestamp
+				host.Hardware.Storage.PhysicalDisks[id] = pd
+			}
+			for _, m := range hostMetadata {
+				if m.Time.Before(time.Now().Add(-timeFilterAge)) {
+					continue
+				}
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: check name before comparing tagsets again.

File: cmd/bosun/sched/host.go [hunk 8]
```diff
@@ -0,0 +1,809 @@
...
+					StatusLastUpdated: timestamp,
+				}
+			}
+		}
+		for _, ts := range hwControllersTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			id, ok := ts["id"]
+			if !ok {
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: perhaps when making maps we could omit datapoints without required tags. That could eliminate a lot of these checks.

File: cmd/bosun/sched/host.go [hunk 9]
```diff
@@ -0,0 +1,809 @@
...
+			if err == nil && timestamp > 0 {
+				c.Status = status
+				c.StatusLastUpdated = timestamp
+				host.Hardware.Storage.Controllers[id] = c
+			}
+			for _, m := range hostMetadata {
+				if m.Time.Before(time.Now().Add(-timeFilterAge)) {
+					continue
+				}
+				if !m.Tags.Equal(ts) {
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: same issue here. let's continue as fast as possible.

File: cmd/bosun/sched/host.go [hunk 10]
```diff
@@ -0,0 +1,809 @@
...
+		var timestamp int64
+		var cpu float64
+		if cpu, timestamp, err = s.Search.GetLast("os.cpu", hostTagSet.String(), true); err != nil {
+			cpu, timestamp, _ = s.Search.GetLast("cisco.cpu", hostTagSet.String(), false)
+		}
+		host.CPU.PercentUsed = cpu
+		host.CPU.StatsLastUpdated = timestamp
+		host.Memory.TotalBytes, timestamp, _ = s.Search.GetLast("os.mem.total", hostTagSet.String(), false)
+		host.Memory.UsedBytes, _, _ = s.Search.GetLast("os.mem.used", hostTagSet.String(), false)
+		host.Memory.StatsLastUpdated = timestamp
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: can you assign `host.Memory.StatsLastUpdated` two lines above inline?

File: cmd/bosun/sched/host.go (hunk 7 above)
Comment (kylebrandt) [inline cmd/bosun/sched/host.go]: In the scenario, where would you recommend I check for the existence of the host in that particular tagset result?

File: cmd/bosun/sched/host.go (hunk 6 above)
Comment (kylebrandt) [inline cmd/bosun/sched/host.go]: In this scenario, where would you recommend I check for the existence of the host in that particular tagset result?

File: cmd/bosun/sched/host.go [hunk 11]
```diff
@@ -0,0 +1,809 @@
...
+				continue
+			}
+			var iface *HostInterface
+			if name := m.Tags["iface"]; name != "" {
+				if host.Interfaces[name] == nil {
+					h := new(HostInterface)
+					host.Interfaces[name] = h
+				}
+				iface = host.Interfaces[name]
+			}
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: if no iface tag, we should probably skip or continue a lot of this loop.

File: cmd/bosun/sched/host.go (hunk 6 above)
Comment (captncraig) [inline cmd/bosun/sched/host.go]: right after you fetch the data. Turn it into a `map[string][]Tagset`. only include tagsets with the tags you will be looking for later.

File: cmd/bosun/sched/host.go (hunk 11 above)
Comment (captncraig) [inline cmd/bosun/sched/host.go]: in any case iface will be nil if there is no iface tag.

File: cmd/bosun/sched/host.go [hunk 12]
```diff
@@ -0,0 +1,809 @@
...
+			if name := m.Tags["iname"]; name != "" {
+				iface.Name = name
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: this will panic if iname without iface ever occurs

File: cmd/bosun/sched/host.go [hunk 13]
```diff
@@ -0,0 +1,809 @@
...
+						case 0:
+							host.VM.PowerState = "poweredOn"
+						case 1:
+							host.VM.PowerState = "poweredOff"
+						case 2:
+							host.VM.PowerState = "suspended"
+						}
+						host.VM.PowerStateLastUpdated = timestamp
+					}
+					//Is this safe?
```
Comment (captncraig) [inline cmd/bosun/sched/host.go]: looks safe enough to me.

File: cmd/bosun/web/web.go [hunk 14]
```diff
@@ -604,11 +604,14 @@ func Last(t miniprofiler.Timer, w http.ResponseWriter, r *http.Request) (interfa
 	if r.FormValue("counter") != "" {
 		counter = true
 	}
-	tags, err := opentsdb.ParseTags(r.FormValue("tagset"))
-	if err != nil {
-		return nil, err
-	}
-	return schedule.Search.GetLast(r.FormValue("metric"), tags, counter)
+	val, timestamp, err := schedule.Search.GetLast(r.FormValue("metric"), r.FormValue("tagset"), counter)
```
Comment (captncraig) [inline cmd/bosun/web/web.go]: no ui depends on this, correct?

**Pull request merged**